### PR TITLE
ci: check certificate for data api endpoint load balancer

### PR DESCRIPTION
**Pull request by sreis**
See https://github.com/opstrace/opstrace/issues/114#issuecomment-743199902

### CODE REVIEW

File: ci/deploy-testremote-teardown.sh [hunk 1]
```diff
@@ -221,35 +221,41 @@ configure_kubectl_aws_or_gcp
...
+            sleep $wait
+        else
+            exit ${retcode}
+        fi
+    done
+}
+
+retry_check_certificate loki.system.${OPSTRACE_CLUSTER_NAME}.opstrace.io:443
+retry_check_certificate cortex.system.${OPSTRACE_CLUSTER_NAME}.opstrace.io:443
+retry_check_certificate system.${OPSTRACE_CLUSTER_NAME}.opstrace.io:443
```
Comment (jgehrcke) [inline ci/deploy-testremote-teardown.sh:258]: :muscle:

**Review by jgehrcke — APPROVED** [review]

**Comment by sreis** [comment]
Well... it's going to fail with the error we are tracking in #114

Data API endpoints are using the correct certificate:
```
2020-12-11T14:35:17Z] + retry_check_certificate loki.system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + count=0
[2020-12-11T14:35:17Z] + retries=3
[2020-12-11T14:35:17Z] + check_certificate loki.system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + openssl x509 -noout -issuer
[2020-12-11T14:35:17Z] + tee /dev/stderr
[2020-12-11T14:35:17Z] + timeout --kill-after=10 10 openssl s_client -showcerts -connect loki.system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + grep 'Fake LE Intermediate'
[2020-12-11T14:35:17Z] depth=1 CN = Fake LE Intermediate X1
[2020-12-11T14:35:17Z] verify error:num=20:unable to get local issuer certificate
[2020-12-11T14:35:17Z] verify return:1
[2020-12-11T14:35:17Z] depth=0 CN = bk-3258-568-a.opstrace.io
[2020-12-11T14:35:17Z] verify return:1
[2020-12-11T14:35:17Z] DONE
[2020-12-11T14:35:17Z] issuer=CN = Fake LE Intermediate X1
[2020-12-11T14:35:17Z] issuer=CN = Fake LE Intermediate X1
[2020-12-11T14:35:17Z] + retry_check_certificate cortex.system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + count=0
[2020-12-11T14:35:17Z] + retries=3
[2020-12-11T14:35:17Z] + check_certificate cortex.system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + openssl x509 -noout -issuer
[2020-12-11T14:35:17Z] + timeout --kill-after=10 10 openssl s_client -showcerts -connect cortex.system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + grep 'Fake LE Intermediate'
[2020-12-11T14:35:17Z] + tee /dev/stderr
[2020-12-11T14:35:17Z] depth=1 CN = Fake LE Intermediate X1
[2020-12-11T14:35:17Z] verify error:num=20:unable to get local issuer certificate
[2020-12-11T14:35:17Z] verify return:1
[2020-12-11T14:35:17Z] depth=0 CN = bk-3258-568-a.opstrace.io
[2020-12-11T14:35:17Z] verify return:1
[2020-12-11T14:35:17Z] DONE
[2020-12-11T14:35:17Z] issuer=CN = Fake LE Intermediate X1
[2020-12-11T14:35:17Z] issuer=CN = Fake LE Intermediate X1
```

UI LB is not ready:
```
[2020-12-11T14:35:17Z] + retry_check_certificate system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + count=0
[2020-12-11T14:35:17Z] + retries=3
[2020-12-11T14:35:17Z] + check_certificate system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + timeout --kill-after=10 10 openssl s_client -showcerts -connect system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:17Z] + tee /dev/stderr
[2020-12-11T14:35:17Z] + openssl x509 -noout -issuer
[2020-12-11T14:35:17Z] + grep 'Fake LE Intermediate'
[2020-12-11T14:35:27Z] unable to load certificate
[2020-12-11T14:35:27Z] 140027065459840:error:0909006C:PEM routines:get_name:no start line:../crypto/pem/pem_lib.c:745:Expecting: TRUSTED CERTIFICATE
[2020-12-11T14:35:27Z] + retcode=1
[2020-12-11T14:35:27Z] + wait=5
[2020-12-11T14:35:27Z] + count=1
[2020-12-11T14:35:27Z] + '[' 1 -lt 3 ']'
[2020-12-11T14:35:27Z] + echo 'failed checking if cluster is using certificate issued by LetsEncrypt, retrying in 5 seconds...'
[2020-12-11T14:35:27Z] failed checking if cluster is using certificate issued by LetsEncrypt, retrying in 5 seconds...
[2020-12-11T14:35:27Z] + sleep 5
[2020-12-11T14:35:32Z] + check_certificate system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:32Z] + timeout --kill-after=10 10 openssl s_client -showcerts -connect system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:32Z] + openssl x509 -noout -issuer
[2020-12-11T14:35:32Z] + tee /dev/stderr
[2020-12-11T14:35:32Z] + grep 'Fake LE Intermediate'
[2020-12-11T14:35:42Z] unable to load certificate
[2020-12-11T14:35:42Z] 139673814086784:error:0909006C:PEM routines:get_name:no start line:../crypto/pem/pem_lib.c:745:Expecting: TRUSTED CERTIFICATE
[2020-12-11T14:35:42Z] + retcode=1
[2020-12-11T14:35:42Z] + wait=5
[2020-12-11T14:35:42Z] + count=2
[2020-12-11T14:35:42Z] + '[' 2 -lt 3 ']'
[2020-12-11T14:35:42Z] + echo 'failed checking if cluster is using certificate issued by LetsEncrypt, retrying in 5 seconds...'
[2020-12-11T14:35:42Z] failed checking if cluster is using certificate issued by LetsEncrypt, retrying in 5 seconds...
[2020-12-11T14:35:42Z] + sleep 5
[2020-12-11T14:35:47Z] + check_certificate system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:47Z] + timeout --kill-after=10 10 openssl s_client -showcerts -connect system.bk-3258-568-a.opstrace.io:443
[2020-12-11T14:35:47Z] + tee /dev/stderr
[2020-12-11T14:35:47Z] + openssl x509 -noout -issuer
[2020-12-11T14:35:47Z] + grep 'Fake LE Intermediate'
[2020-12-11T14:35:57Z] unable to load certificate
[2020-12-11T14:35:57Z] 139635737265280:error:0909006C:PEM routines:get_name:no start line:../crypto/pem/pem_lib.c:745:Expecting: TRUSTED CERTIFICATE
[2020-12-11T14:35:57Z] + retcode=1
[2020-12-11T14:35:57Z] + wait=5
[2020-12-11T14:35:57Z] + count=3
[2020-12-11T14:35:57Z] + '[' 3 -lt 3 ']'
[2020-12-11T14:35:57Z] + exit 1
[2020-12-11T14:35:57Z] + teardown
[2020-12-11T14:35:57Z] + LAST_EXITCODE_BEFORE_TEARDOWN=1
```

**Comment by sreis** [comment]
During the install, it took almost two minutes for the data API endpoints to return a valid response:

```
[2020-12-11T14:33:12Z] 2020-12-11T14:33:12.102Z info: waiting for expected HTTP responses at these URLs: {
```

```
[2020-12-11T14:35:15Z] 2020-12-11T14:35:15.923Z info: All probe URLs returned expected HTTP responses, continue
```

**Comment by sreis** [comment]
I'm adding a quick check in the installer to wait for grafana. Let's see what the behavior is...

**Comment by sreis** [comment]
It looks like this:
```
[2020-12-11T18:13:09Z] 2020-12-11T18:13:09.203Z info: waiting for expected HTTP responses at these URLs: {
[2020-12-11T18:13:09Z]   "https://default.bk-3268-640-a.opstrace.io/": "default",
[2020-12-11T18:13:09Z]   "https://system.bk-3268-640-a.opstrace.io/": "system"
[2020-12-11T18:13:09Z] }
[2020-12-11T18:13:09Z] 2020-12-11T18:13:09.370Z info: https://system.bk-3268-640-a.opstrace.io/: got expected HTTP response
[2020-12-11T18:13:09Z] 2020-12-11T18:13:09.373Z info: https://default.bk-3268-640-a.opstrace.io/: got expected HTTP response
```

Both CI runs didn't trigger the flake, so I'll merge to, hopefully, get a signal during the weekend. We can polish it afterward.

**Pull request merged**